pocket_coffea/lib/delayed_eval.py

```diff
@@ -0,0 +1,128 @@
+            }
+        else:
+            self.registered[name] = {
+                "compute_fn": compute_fn,
+                "default": default_value,
+                "multi": False,
+            }
+
+    def ensure_nominal_index(self, events):
+        """Attach a stable nominal index column used for vectorized mapping across variations."""
+        idx_name = "__poco_nom_idx_delayed_eval__"
+        if idx_name not in ak.fields(events):
+            # Use a simple per-event counter on the outer axis; ensures stable mapping across variations
+            events[idx_name] = ak.Array(np.arange(len(events), dtype=np.int64))
+        return idx_name
+
+    def prepare_nominal_snapshot(self, events):
+        if self.prepared or len(self.registered) == 0:
+            return
+        self.nominal_events = events
+        # Ensure a persistent nominal index column is present
+        self.nom_idx_field = self.ensure_nominal_index(events)
+        n = len(events)
+        # Per-branch caches and bookkeeping
+        self.cache = {}
+        self.computed_nom_mask = np.zeros(n, dtype=bool)
+        for key, cfg in self.registered.items():
+            if cfg.get("multi", False):
+                for nm in cfg["names"]:
+                    self.cache[nm] = np.full(n, cfg["default"], dtype=float)
+            else:
+                self.cache[key] = np.full(n, cfg["default"], dtype=float)
+        self.prepared = True
+
+    def union_final_categories_mask(self, events, categories):
+        # Build union-of-final-category mask using the categories masks
+        union = events.event != events.event  # False
+        for _, mask in categories.get_masks():
+            if getattr(categories, "is_multidim", False) and getattr(mask, "ndim", 1) > 1:
+                mask_on_events = ak.any(mask, axis=1)
+            else:
+                mask_on_events = mask
+            union = np.logical_or(union, mask_on_events)
+        return union
+
+    def update_for_current_variation(self, events, categories):
+        if not self.prepared or len(self.registered) == 0:
+            return
+
+        # Which current events pass final selections in this variation
+        passing_any = self.union_final_categories_mask(events, categories)
+        # Nominal indices for the current (filtered) events are carried along as a field
+        idxs_curr = ak.to_numpy(events[self.nom_idx_field])
+        valid = idxs_curr >= 0
+
+        # Newly passing nominal indices
+        passing_idxs = idxs_curr[passing_any]
+        passing_valid = passing_idxs >= 0
+        nom_idxs_pass = passing_idxs[passing_valid]
+        if nom_idxs_pass.size > 0:
+            # Filter to not-yet-computed and unique indices
+            not_done_mask = ~self.computed_nom_mask[nom_idxs_pass]
+            nom_idxs_new = np.unique(nom_idxs_pass[not_done_mask])
+            if nom_idxs_new.size > 0:
+                ev_sub = self.nominal_events[nom_idxs_new]
```

Comment:
HI! Can you please guide me a bit about what's happening in this section? I'm trying to understand why the processor object is needed on the delayedeval class side